feat: working simulation in cyberbook

heaps0rt opened the pull request:
Working simulation package which works as intended in cyberbook

heaps0rt committed:
feat: Outputs the result of the simulation which can be used for evaluation of questions related to the simulation

Modified the widgets.py file to make the result and timestep of the simulation global so that it is accessible for the evaluation.

heaps0rt committed:
fix: fixed solara not being able to be installed

Ignores all mentions of solara in the code for micropip to install the ipysim package.

heaps0rt committed:
fix: Better handling of slider values and documentation
Fixes certain values generating simulations that are not able to render. In addition, widgets.py has been documented to a greater extent.

heaps0rt committed:
fix: More concise documentation
Documentation is not now more, and changes meant for another issue in another branch have been reverted

heaps0rt committed:
fix: Better handling of values for simulation

heaps0rt committed:
feat: Rudimentary evaluation of questions related to simulation
A basic implementation of question evaluation. Teachers can define values for Kp and Kd that are correct.

heaps0rt committed:
correct notebook

jorgengaldal committed:
feat: add evalution_function

jorgengaldal committed:
fix: small example

heaps0rt committed:
fix: remove scipy warning suppress

heaps0rt committed:
chore: removed pycache from repo

jorgengaldal commented on src/ipysim/widgets.py:
```diff
@@ -26,54 +33,120 @@ def interactive_simulation(
+        dt (float): Time step for the simulation.
+        Kp_default (float): Default proportional gain for the PD controller.
+        Kd_default (float): Default derivative gain for the PD controller.
+        target_kp (float): Target proportional gain for evaluation.
+        target_kd (float): Target derivative gain for evaluation.
```
```suggestion

```

heaps0rt committed:
Update src/ipysim/widgets.py

Co-authored-by: Jørgen Galdal <jorgen.4@live.no>

jorgengaldal commented on src/ipysim/widgets.py:
```diff
@@ -26,54 +33,120 @@ def interactive_simulation(
+        except Exception as e:
+            with out:
+                out.clear_output(wait=True)
+                print(f"Error: {e}")
```
Bedre error handling her for brukere? Ikke bare printe feilen som ikke gir mening for brukere, kanskje?
```suggestion
            with out:
                out.clear_output(wait=True)
                print("An error occurred. Please try again.")
```

heaps0rt committed:
feat: Added improved error handling
Error message will now be more informative and values will be rolled back if the simulation fails

heaps0rt committed:
Merge branch 'evaluation' of github.com:forceofthecyber/ipysim into evaluation

heaps0rt committed:
feat: Added improved error handling
Informs the user on when the simulation fails and rolls back to previous valid values

heaps0rt committed:
fix: remove unnecessary warnings and logging to console
ipysim now removes all warnings and has proper error handling for the user which is informative. All errors are rerouted to the console of the browser.

jorgengaldal reviewed (approved):
Ser bra ut! Ser ut til å fungere bra i boken også.
